xyzservices/lib.py

```diff
@@ -28,6 +28,70 @@ def __call__(self, **kwargs):
         new.update(kwargs)
         return new
 
+    def build_url(self, x=None, y=None, z=None, token=None, scale_factor=None):
+        """
+        Build the URL of tiles from the TileProvider object
+
+        Can return URL with placeholders or the final tile URL.
+
+        Attributes
+        ----------
+
+        x, y, z : int
+            tile number
+        token : str (optional)
+            Access token (or API key or similar) for the tiles requiring one.
+        scale_factor : str (optional)
+            Scale factor (where supported). For example, you can get double resolution
+            (512 x 512) instead of standard one (256 x 256) with ``"@2x"``. If you want
+            to keep a placeholder, pass `"{r}"`.
+
+        Returns
+        -------
+
+        url : str
+
+        Examples
+        --------
+        >>> import xyzservices.providers as xyz
+
+        >>> xyz.CartoDB.DarkMatter.build_url()
+        'https://a.basemaps.cartocdn.com/dark_all/{z}/{x}/{y}.png'
+
+        >>> xyz.CartoDB.DarkMatter.build_url(x=9, y=11, z=5)
+        'https://a.basemaps.cartocdn.com/dark_all/5/9/11.png'
+
+        >>> xyz.CartoDB.DarkMatter.build_url(x=9, y=11, z=5, scale_factor="@2x")
+        'https://a.basemaps.cartocdn.com/dark_all/5/9/11@2x.png'
+
+        >>> xyz.MapBox.build_url(token="my_token")
+        'https://api.mapbox.com/styles/v1/mapbox/streets-v11/tiles/{z}/{x}/{y}?access_token=my_token'
+
+        """
+        provider = self.copy()
+
+        if not x:
+            x = "{x}"
+        if not y:
+            y = "{y}"
+        if not z:
+            z = "{z}"
+
+        if token:
+            for key, val in provider.items():
+                if isinstance(val, str) and "<insert your" in val:
+                    provider[key] = token
```

Comment:
I am not sure this is fully robust. There are tile providers where there are 2 items to fill in (eg app_id and app_code for HERE). 

Generally the explicit way is that the user provides the key / token creating an updated provider object with those values filled in before passing it to the package using the object (contextily, folium, ..). 
It might still be worth having a user-friendly shortcut keyword (that is always the same) for the cases where it can work, but not fully sure about this.

Reply:
Something we could also do is maybe raise an error/warning if at this point there is still a field with "insert your ..", to avoid building a url with that text (which I think will always be wrong / unintentional?)

Reply:
Hmm, I think you're right. Warning may be the best solution.